File: channel_select.m
function ch_selected = channel_select(A, b, N, varargin)

% channel_select A function which selects the best N channels of A for 
%                the least-squared problem min_{x} 1/2 * ||Ax - b ||² 
%                which best estimates b.
%
%  ch_selected = channel_select(A, b, N, 'method', 'utility', 'group_size', group_size)
%                Selects the best N channels of A using greedy 
%                utility-based channel selection. Here the columns of A are
%                assumed to contain groups of group_size channels in consecutive columns.
%
%  ch_selected = channel_select(A, b, N, 'method', 'utility', 'grpid',  grpid)
%                Selects the best N group-IDs of A using greedy 
%                utility-based channel selection. Here the columns of A are
%                assumed to contain groups of channels in non-consecutive columns.
%                Each column of A belongs to a group corresponding to its index 
%                in grpid.
% Required Inputs:
%   A - Data/measurements used in estimation where P is the number of 
%       channels (including delayed versions) and T is the number of
%       time samples. 
%       If A consists of delayed samples of a channel, all delays of 
%       a channel are expected to be consecutive columns
%       Size: (T X P)
%   b - The desired data to be estimated using A. Size: (T X 1)
%   N - Number of channels to be selected.
% Optional Inputs (Name-value pairs: Each optional input string is to be followed by an argument):
%   'method' - A string for the method to be used to select best N channels
%            Currently supported:
%            'utility' - Utility based channel reduction. Here, if A is 
%            singular, utility definition with minimum norm increase 
%            interpretation is used.
%            (ref. A. Bertrand, 2018)
%
%   'group_size' - Number of columns of A to be considered as a group.
%          For eg: Channel-k and its time-lagged copies can be together considered as
%          a group. The channels in the group should be in consecutive
%          columns.
%          The utility of this group is considered as the utility of channel-k.
%          If every channel and their 2 time-lagged versions (a channel itself + 2 lagged versions) are present, then:
%          (group size in the above case would be 3.)
%          default: 0 (no time-lagged versions included. group size is 1)
%  'grpid' - A vector of length equal to number of columns. The vector contains integer entries.
%	     Each integer entry corresponds to the ID of a column. All columns belonging to a group should correspond to same ID.
%	     Every column should have an ID.  This argument allows custom arrangement of columns belonging to same group.
%	     NOTE: This argument is ignored if 'group_size' input is present. 
%               
%   
%
% Outputs: 
%   ch_selected - The best N columns in A or the best N groups of A based 
%                 on group-IDs. 
%                 There is NO ORDER OF SIGNIFICANCE for this list.
%
% author: Abhijith Mundanad Narayanan
% e-mail: mailme@abhimundanad.com
%
%

    % Default conditions
    method = 'utility';
    noflags = 1;
    lagsflag = 1;
    if(nargin<3)
        N = size(A,2);
    end
    
    
    if nargin > 3        
        for i = 1:2:length(varargin)
            Param = varargin{i};
            Value = varargin{i+1};
            if ~isstr(Param)
                error('Flag arguments must be strings')
            end
            Param = lower(Param);
            switch Param
                case 'method'
                    method = Value;
		% The following case is for backward compatibility for an earlier version of this
		% code which had 'lags' argument instead of 'group_size'
                case 'lags'
                    noflags = Value+1;
                    lagsflag = 1;
                case 'group_size'
                    noflags = Value;
                    lagsflag = 1;
                case 'groupid'
                    grpid = Value;
                    nofuniq = unique(grpid); % Number of unique groups
                    lagsflag = 0;
                    if(size(grpid,1)~=size(A,2))
                        error('Length of group-ids != No.of columns in A');
                    end
            end
        end
    end
    
    if(strcmp(method,'utility'))
        % Calculate the auto and crosscovariances
        RXX = (A'*A)/size(A,1);
        RXY = (A'*b)/size(A,1);
        no_of_channels = size(RXX,2)/noflags;

        % Initialise a list of original indices/channels
        chnl_list = (1:no_of_channels)';

        % Indicator matrices to book-keep removed channels

        % selector of columns (channels and lags)
        col_sel = ones(1,size(RXX,2));

        % indicator vector for channels
        node_ids = ones(1,no_of_channels);
        node_ids = logical(node_ids);

        % Vector to store removed channels
        deleted_channels = zeros(no_of_channels,1);
        del_ch_count = 0;
        indep_vars = 0;
        recursive_comp = 0;
        csflag = 1;

        while(csflag)

            % Populate a list of remaining channel numbers
            temp_chnl_list = chnl_list(node_ids,:);
            X_sel = RXX(logical(col_sel), logical(col_sel));
            RXY_sel = RXY(logical(col_sel),:);              

            % Check if normal utility or minimum norm utility definition is
            % to be used based on the covariance being singular
            if(rcond(X_sel)<1e-12)
                min_norm_flag = 1;
            else                
                min_norm_flag = 0;
                % Keeps track of independent columns/variables in the data
                indep_vars = indep_vars+1;
            end
            
            util = zeros(size(temp_chnl_list,1),1);
            
            if(min_norm_flag)                         
                 % Recursive computation of inverse - reduce complexity
                 % Only after the first inverse has been computed
                if(recursive_comp)
                    k = idx;                                       
                    S = Xinv((k-1)*noflags+1:k*noflags,(k-1)*noflags+1:k*noflags);
                    
                    V1 =  Xinv((k-1)*noflags+1:k*noflags,1:(k-1)*noflags);
                    V2 =  Xinv((k-1)*noflags+1:k*noflags,(k*noflags)+1:end);
                    V = [V1,V2];
                    
                    C = [];
                    C1 = Xinv(1:(k-1)*noflags,1:(k-1)*noflags);
                    C = [C1,Xinv(1:(k-1)*noflags,(k*noflags)+1:end)];
                    C2 = Xinv((k*noflags)+1:end,1:(k-1)*noflags);
                    C = [C;C2,Xinv((k*noflags)+1:end,(k*noflags)+1:end)];
                    Xinvnew = C- V'*(S\V);
                    
                    Xinv = Xinvnew;
                else
                    % Find regularized inverse for minimum norm utility definition
                    % Select the minimum positive eigenvalue as lambda
                    eigvals = eig(X_sel);
                    lambda_scaling = min(eigvals(eigvals>1e-4));
                    lambda_I = (lambda_scaling*1.0e-5)*eye(size(X_sel,1)); 
                    Xinv = (X_sel + min_norm_flag*lambda_I)\eye(size(X_sel,1));
                    recursive_comp = 1;
                end
            else
                Xinv = X_sel\eye(size(X_sel,1));
            end

            % Compute the new decoder with remaining channels
            W = Xinv * RXY_sel;
            
            % Create group-IDs if lagged versions in data
            if(lagsflag)
                nofgrps = size(temp_chnl_list,1); 
                grpid = repmat(1:nofgrps, noflags, 1);
                grpid = reshape(grpid(:,1:end),[],1);
            end
            % Compute utility of all remaining channels
%             util = grputilcalc(Xinv, W, grpid);
            
            for k = 1:size(temp_chnl_list,1)
                % Select decoder weights of channel k and its lags
                Wkq = W((k-1)*noflags+1:k*noflags);
                % utility computation of channel k ( if block utility: channel and its lags)
                S = Xinv((k-1)*noflags+1:k*noflags,(k-1)*noflags+1:k*noflags);
                util(k) = Wkq'*(S\Wkq);
            end

            % Find the index of channel with least utility
            [~, idx] = min(util);

            % Pick the actual channel number of least utility
            % from list of remaining channel numbers
            temp_ch_sel = temp_chnl_list(idx, :);

            % Delete that channel from the set
            row_id = find(chnl_list==temp_ch_sel);
            col_sel((row_id-1)*noflags+1:row_id*noflags) = 0;       
            node_ids(row_id) = false;

            % store the deleted/removed channel
            del_ch_count = del_ch_count+1;            
            deleted_channels(del_ch_count,1) = temp_ch_sel;                
            
            if(no_of_channels-del_ch_count==N || del_ch_count==N)
                csflag = 0;
            end
        end

        % Best N channels/group-IDs in ascending order of significance
        % i.e. the last deleted channel comes first        
        % Storing deleted channels to give an order of significance if
        % selecting all channels
        
        if(no_of_channels==N)
            ch_selected = flipud(deleted_channels);
            ch_selected = ch_selected(1:N);        
        % Best N channels
        else
            ch_selected = chnl_list(node_ids,:);
        end
    else
        error('Supported Methods: utility');
    end
        
end

File: channel_select.py
import numpy as np
from itertools import compress
def channel_select(**argV):
''' 
A function which selects the best columns of A in the least-squares problem 
  min_x 1/2 ||Ax-b||_2^2 

  Usage: ch_selected = channel_select(A=A, b=b, method='utility')
  
  Input: A - matrix with channels/features along columns and time/length of features along rows
            size: (T X N) - N channels; T duration/length
        b - Vector of size (T X 1)
         method - a string; currently supported : 'utility' [1,2]
 
 [1] Narayanan, A. M., & Bertrand, A. (2019). Analysis of miniaturization effects and channel selection strategies for EEG sensor networks with application to auditory attention detection. IEEE Transactions on Biomedical Engineering.
 [2] Bertrand, A. (2018). Utility Metrics for Assessment and Subset Selection of Input Variables for Linear Estimation [Tips & Tricks]. IEEE Signal Processing Magazine, 35(6), 93–99.


'''
   if 'A' in argV:
        A = argV['A']
    if 'b' in argV:
        b = argV['b']
    if 'N' in argV:
        N = argV['N']        
    else:
        print('Number of channels to be selected required!')
    if 'method' in argV:
        method = argV['method']
    else:
        # By default channel selection method is utility
        method = 'utility'
    if 'lags' in argV:
        lags = argV['lags'] + 1
    else:
        lags = 1
    
    if method is 'utility':
        # Get the dimensions of the data
        m = A.shape[0]
        n = A.shape[1]
        
        
        no_of_channels = int(n/lags)
        ch_list = list(range(1,no_of_channels+1))
        
        # Compute covariances
        At = np.matrix.transpose(A)
        AtA = At @ A
        Atb = At @ b
        
        # Initialise logical indexing masks to select/remove channels/columns 
        utility_vals = np.zeros((no_of_channels,1), dtype=float)
        del_ch_count = 0
        col_sel_mask = np.ones((n,), dtype=bool)
        ch_sel_mask = np.ones((no_of_channels,), dtype=bool)
        ch_sel = list(compress(ch_list,ch_sel_mask))
        ch_selected = []

        # Start iterative channel deletion based on utility
        while del_ch_count<no_of_channels:
            
            AtA_sel = AtA[:,np.array(col_sel_mask)]
            AtA_sel = AtA_sel[np.array(col_sel_mask),:] 
            
            
            
            # Estimate the current decoder using (A'A)^(-1)(A'b)
            
            # This decoder is used for the fast computation of utlity of a column or a group of columns
            AtA_rank = np.linalg.matrix_rank(AtA_sel)
            if AtA_rank==AtA_sel.shape[1]:
                # If AtA is full-rank, normal definition of utility
                AtA_sel_inv = np.linalg.inv(AtA_sel)
            else:                
                # l2-regularization based definition of utility when AtA is not full-rank
                id_mat = np.identity(AtA_sel.shape[1])
                diag_mat = np.diag(AtA_sel)
                nz_diag = diag_mat>0
                lmb_val = np.amin(diag_mat[nz_diag])
                AtA_sel = AtA_sel + (lmb_val * 1e-5 * id_mat)
                AtA_sel_inv = np.linalg.inv(AtA_sel)
            Atb_sel = Atb[col_sel_mask]
            W_sel = AtA_sel_inv @ Atb_sel            
            indx = np.array(np.where(ch_sel_mask==True))
            util = np.zeros((indx.shape[1],1),dtype='float')
            k = 1      
        # Compute (group)-utility of columns of current A
            while k<=len(util):
                if lags==1:
                    S = AtA_sel_inv[k-1][k-1]
                    util[k-1] = np.matrix.transpose(W_sel[k-1]) @ ((1/S) * W_sel[k-1])
                else:
                    strt = (k-1)*lags
                    stp = strt+lags
                    S = AtA_sel_inv[strt:stp,:]
                    S = S[:,strt:stp]
                    W_block = W_sel[strt:stp]
                    util[k-1] = np.matrix.transpose(W_block) @ np.linalg.inv(S) @ W_block
                k = k+1
                
        # Eliminate the column or group of columns corresponding to lowest utility
            min_util = np.argmin(util)
            ch_sel_mask[ch_sel[min_util]-1] = False
            strt = (ch_sel[min_util]-1)*lags
            stp = strt+lags
            col_sel_mask[strt:stp] = False
       # Append the deleted column number to a list
            ch_selected.append(ch_sel[min_util])
            ch_sel = list(compress(ch_list,ch_sel_mask))
            del_ch_count = del_ch_count+1
      # The order of channels in ch_selected list in the reverse of their importance
      # That is, the first channel in the list is the least important (one that was deleted first)
        return [ch_selected]
            
    return
